CDTDatastore/HTTP/CDTIAMSessionCookieInterceptor.m

```diff
@@ -0,0 +1,145 @@
+//
+//  CDTIAMSessionCookieInterceptor.m
+//  CDTDatastore
+//
+//  Created by tomblench on 05/07/2017.
+//  Copyright © 2017 IBM Corporation. All rights reserved.
+//
+
+#import <Foundation/Foundation.h>
+
+#import "CDTIAMSessionCookieInterceptor.h"
+#import "CDTLogging.h"
+
+static const NSInteger CDTIAMSessionCookieRequestTimeout = 600;
+
+
+@interface CDTIAMSessionCookieInterceptor ()
+
+
+@property NSData *iamSessionRequestBody;
+
+- (NSData *)getBearerToken;
+
+/** NSURLSession to make calls to get IAM bearer token (shouldn't be same one we're intercepting). */
+@property (nonnull, nonatomic, strong) NSURLSession *iamUrlSession;
+
+@property NSURL *iamTokenUrl;
+
+@end
+
+@implementation CDTIAMSessionCookieInterceptor
+
+- (instancetype)initWithAPIKey:(NSString *)apiKey
+{
+    self = [super init];
+    if (self) {
+        // build the request to get the IAM bearer token
+        _iamSessionRequestBody = [[NSString stringWithFormat:@"grant_type=urn:ibm:params:oauth:grant-type:apikey&response_type=cloud_iam&apikey=%@", apiKey] dataUsingEncoding:NSUTF8StringEncoding];
+        
+        NSURLSessionConfiguration *config =
+        [NSURLSessionConfiguration ephemeralSessionConfiguration];
+        // The IAM bearer token endpoint requires a form-encoded request with API key present.
+        // We might as well set that up now.
+        config.HTTPAdditionalHeaders = @{ @"Content-Type" : @"application/x-www-form-urlencoded" };
+        _iamUrlSession = [NSURLSession sessionWithConfiguration:config];
+        [super setShouldMakeSessionRequest:YES];
+        _iamTokenUrl = [NSURL URLWithString:@"https://iam.bluemix.net/oidc/token"];
+    }
+    return self;
+}
+
+- (NSURLSessionConfiguration*)customiseSessionConfig:(NSURLSessionConfiguration*)config
+{
+    // set header for cookie session
+    // The _iam_session endpoint requires an application/json body.
```

Comment:
`requires` is slightly misleading, since we choose `application/json` over form encoding